### DISCUSSION TITLE
Export annotation for range limits

### DISCUSSION THREAD

**Discussion by AndreaTerenz** [godotengine/godot-proposals#10545] [Scripting]
In a `ParticleProcessMaterial` you can find many parameters that are randomly set, for each particle, to a value within a user defined range (e.g., radial acceleration or initial velocity). This is done with a UI that "visualizes" the range's limits and very easily lets you change them:
![image](https://github.com/user-attachments/assets/db0c5fe3-b279-4a94-9138-78eed2485e7b)

Unlike any other UI element in the editor (as far as I can tell), this is _not_ possible anywhere else, since these sliders are of a purpose-built [ParticleProcessMaterialMinMaxPropertyEditor](https://github.com/godotengine/godot/blob/a1acd38c35781181cde726b039eb8a9a0b6aa79f/editor/plugins/particle_process_material_editor_plugin.h#L44) class, not used anywhere else. This also means that there is no way to `@export` a pair of variables with such a UI from a script. I propose to change that.

The idea is to add a `@export_minmax` annotation, which would show that UI in the editor and be used, in code, like this:

```gdscript
# other names could be @export_subrange or @export_range_limits
@export_minmax(hard_min, hard_max, step)
var minimum := default_minimum
var maximum := default_maximum
```

This annotation would require to be followed by two int or float variables of the same type. The  "step" parameter would be optional and change the granularity of the sliders (just like with `@export_range`).

Alternatively, to maintain that export annotations should be about just one variable, this could be done through `Vector2i` (for `int` limits) or `Vector2` (for `float` limits)

```gdscript
@export_minmax(hard_min, hard_max, step) var minmax_float : Vector2 = some_default
@export_minmax(hard_min, hard_max, step) var minmax_int : Vector2i = some_int_default
```

Though maybe it would be awkward to think of range limits as xy components of a vector

**Comment by AndreaTerenz** [comment]
PS: to be clear, my point is about easily editing the limits of a range. Whether those are used to randomly set a thid variable, like in a Particle Material, is beside the point

**Comment by Pshy0** [comment]
**Alternative:**
Godot presents different property editing UI depending on the data type.
Perhaps this can be done by providing a data type representing a min/max range.
Then no new annotation would be needed, `@export_range` could just be used:
```gdscript
@export_range(hard_min, hard_max, step) var minmax_float : MinmaxVector2 = some_default
@export_range(hard_min, hard_max, step) var minmax_int : MinmaxVector2i = some_int_default
```
`MinmaxVector2` and `MinmaxVector2i` (arbitrarily named here, it can be something else) would have a `min` and `max` properties, instead of `x` and `y`, so would have the advantage of being less confusing than using `Vector2` and `Vector2i`.

**Reply by AndreaTerenz** [reply]
A better name could be something like `Range` ~or `Span`~ (and thus `IntRange`/`Rangei`~/`IntSpan`/`Spani`~), which would "free" this from having to pretend to be a vector, having additional getter attributes such as `center` and `width`

With other types, `@export_range`'s UI would behave as usual, while with this new type it would show the aforementioned UI

It could also help in general by making it slightly easier to pass a range of value around as a single variable instead of pairs of min-value/max-value

EDIT: `Span` is probably a bad idea since it's the name of a completely different thing already in C#

**Reply by penggrin12** [reply]
`Range` is also taken by a `Control` class that ProgressBars inherit from.

**Reply by AndreaTerenz** [reply]
@penggrin12 crikey that's true
Maybe `Interval` then?
